### PR TITLE
[ml/air] Checkpoints serialization/deserialization support

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This PR adds support for checkpoint ser/de. In particular this is special casing the local data representation, which will be converted into a bytes checkpoint on serialization. This way checkpoint objects sent to remote tasks are guaranteed to always point to a valid data location within the remote task.
We are not detecting pickling to/from disk (e.g. to pickle files) for now.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/checkpoint.py [hunk 1]
```diff
@@ -99,6 +99,21 @@ class Checkpoint:
             # It is guaranteed that the original data was recovered
             assert isinstance(clf, RandomForestClassifier)
 
+        Checkpoints can be pickled and sent to remote processes.
+        Please note that checkpoints pointing to local directories will be
+        pickled as data representations, so the full checkpoint data will be
+        contained in the checkpoint object. If you want to avoid this,
+        consider passing only the checkpoint directory to the remote task
+        and re-construct your checkpoint object in that function. Note that
+        this will only work if the "remote" task is scheduled on the
```
Comment (xwjiang2010) [inline python/ray/ml/checkpoint.py:108]: Slightly off the topic as we are mostly talking about using checkpointing across different libraries. But do you foresee us also using this inside of Tune? For example, trial1 saves a checkpoint on node1, which needs to be restored and mutated onto trial2 on node2. In this pbt case, can the driver code be something like 
checkpoint1 = ray_trial_executor.save(trial1) # "serialized" form that basically says you can find it on this path on node2
ray_trial_executor.mutate(trial2, checkpoint1, some_mutation) # which is allocated on node2, which we have no control of.
Do you think Checkpoint impl will be responsible for rsyncing that file from node1 to node2, instead of just throwing errors?
Basically the question is about efficiency and usability - if rsync is more efficient than object store, can user expect Checkpoint impl to rsync files onto different nodes?

**Review by ericl — APPROVED** [review]

File: python/ray/ml/checkpoint.py (hunk 1 above)
Comment (ericl) [reply python/ray/ml/checkpoint.py:108]: > Do you think Checkpoint impl will be responsible for rsyncing that file from node1 to node2, instead of just throwing errors?

I don't think Tune should rely on this--- it should directly manipulate the object ref / etc. data in CheckpointManager. The Checkpoint class should only be added at the last minute, when returning the results to the user.
